### ISSUE TITLE
Snippet / Modeladmin / settings forms should use `validation_error` helper

### ISSUE THREAD

**Issue by gasman** [wagtail/wagtail#3441]
See #3440, which introduces a new `validation_error` helper to ensure that validation errors on the page edit form - both field-specific, and non-field errors - are displayed appropriately. After this is merged, we should update the other admin views that work with user-defined forms so that they're also using this helper:

* snippets
* ModelAdmin
* `wagtail.contrib.settings`

As part of this task, we should make sure that validation error handling is covered by unit tests - i.e. for each of these modules, there should be at least one test similar to https://github.com/wagtail/wagtail/pull/3440/files#diff-509186d17133395aafeb7c62930e9881R3964 which posts an invalid form and confirms that the form is redisplayed with the expected error message included. (There's no need to go through all permutations of field and non-field errors, since that's already covered by #3440.)

**Comment by ababic** [comment]
Hi @gasman, do you know if the solution implemented for pages / proposed here will surface non-field errors from `InlinePanel` instances too?

**Comment by gasman** [comment]
@ababic I suspect it won't... [`ClusterForm.is_valid`](https://github.com/wagtail/django-modelcluster/blob/fa1e2fc3194c8676e3ae5a2beeb7de25c6d8f219/modelcluster/forms.py#L258) recursively checks the `is_valid` of the associated formsets, so it will at least recognise that validation has failed. However, to actually report the errors we'd have to make the `errors` dict of the child forms/formsets bubble up to the top level (or alternatively, have the `validation_error` helper dig in to `form.formsets['foo'].errors`, but that seems like a bad idea - `validation_error` should ideally not have any special case logic for ClusterForm, and certainly shouldn't have to interface with it at such a low level).

I guess the correct behaviour for "how should errors in child formsets be surfaced in `form.errors`" has never been defined - Django doesn't have an answer to that, because child formsets are a modelcluster-specific concept - so that would be the first thing to address. Two possibilities spring to mind:

- extend the error dict object to have a `CHILD_FORM_ERRORS` property alongside `NON_FIELD_ERRORS` and the per-field keys, something like:

        event_form.errors = {
            'title': ["This field is required"],
            'NON_FIELD_ERRORS': ["End date must be after start date"],
            'CHILD_FORM_ERRORS': {
                'speakers': {
                    0: {'NON_FIELD_ERRORS': "Speaker must agree to the T+Cs"}
                }
        }

- treat formsets as equivalent to fields of the top-level form (since that's basically what happens when accessing the ClusterForm object itself) - this way we're not customising the error dict API, so any existing code that talks to the error dict (e.g. the `validation_error` helper) will continue to just work:

        event_form.errors = {
            'title': ["This field is required"],
            'NON_FIELD_ERRORS': ["End date must be after start date"],
            'speakers': ["Speaker must agree to the T+Cs"],
        }

    I'm not sure what we do about field errors within the child forms (e.g. "Speaker name is required"), though...

**Comment by ababic** [comment]
Thanks @gasman for the full and prompt reply. Having reviewed the changes in #3440 more fully, and considering the above, I have a few thoughts. 

Do child form errors need to be surfaced so fully in `validation_error` for the parent? I feel like it could be enough to render a generic error message about child form errors at the parent level, such as:

`One/two/three/several speakers are not valid`

Then each child form would have it's own `validation_error` to combine / display errors for that form in the same fashion as the parent.

Structurally,  I suppose the parent form error structure could look something like:

```
   event_form.errors = {
      'title': ["This field is required"],
      'NON_FIELD_ERRORS': ["End date must be after start date"],
      'CHILD_FORM_ERRORS':  {
          'speakers': [0, 2, 5]   # indexes of invalid child forms / fieldsets,
      }
```

This does feel like we're making a special case for `ClusterForm`, but I think that's only because a `ClusterForm` / `InlinePanel` combination is currently the only way in wagtail to define interfaces that suit related objects being defined together in such a way. But, that is really only a limitation for the time being. Conceptually speaking, any form should be able to have children forms (even children), so having a `validation_error` implementation to support that consistently makes sense.

**Comment by gasman** [comment]
@ababic Good point - I think we only need a generic message in the top-level `validation_error`, since the more specific error can be shown inline on the child object. (In the case of a `NON_FIELD_ERROR` on the child object this probably needs some new code, but that's doable.) In that case, do we really need any special-casing on the top-level error dict? It seems to me that we can just say:

    event_form.errors = {
      'title': ["This field is required"],
      'speakers': ["Three speakers are not valid"],
      'NON_FIELD_ERRORS': ["End date must be after start date"],
   }

**Comment by ababic** [comment]
Thanks @gasman. I do like the simplicity of that. However, mixing them in with bona-fide `Field` errors  doesn't feel as pythonic to me. I imagined `validation_error` would do the string representation bit, but the error dict itself could provide something more useful (For example, it would be possible to create a custom `validation_error` that contained same-page anchor links to the invalid children)

**Comment by ababic** [comment]
I feel I may have been a little vague above. To clarify, when I said:

> However, mixing them in with bona-fide Field errors doesn't feel very pythonic to me

What I meant was, I feel like `errors` is really data for developers, so distinguishing between different types of errors there provides fuller picture of what is happening, and reduces ambiguity about where the values came from. `validation_error` exists to render consistently formatted errors for end users, so to me, that feels like a more natural place to generate the `"Three speakers are not valid"` string.
